src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -1054,8 +1054,15 @@ ExecuteModifyTasks(List *taskList, bool expectResults, ParamListInfo paramListIn
 		LockPartitionRelations(firstShardInterval->relationId, RowExclusiveLock);
 	}
 
-	/* ensure that there are no concurrent modifications on the same shards */
-	AcquireExecutorMultiShardLocks(taskList);
+	/*
+	 * Ensure that there are no concurrent modifications on the same
+	 * shards. For DDL commands, we already obtained the appropriate
+	 * locks in ProcessUtility.
+	 */
+	if (firstTask->taskType != DDL_TASK)
```

Comment:
Maybe `taskType == MODIFY_TASK`? 

Might be worth mentioning that the current motivation for `AcquireExecutorMultiShardLocks` is two-fold:
- To prevent multi-shard UPDATE/DELETE/INSERT...SELECT from running concurrently with each other. This is not strictly necessary but it avoids making an excessive number of connections and creating many distributed deadlocks.
- Ensure that non-commutative writes are serialised when replication factor > 1

Reply:
> Might be worth mentioning that the current motivation for AcquireExecutorMultiShardLocks is two-fold:

Do you mean here in the comment? We already have explained these in comments of `AcquireExecutorMultiShardLocks`.